src/MonitorProfiler/MainProfiler/ExceptionTracker.cpp

```diff
@@ -0,0 +1,304 @@
+
+    HRESULT hr = S_OK;
+
+    IfFailLogRet(_threadDataManager->SetHasException(threadId));
+
+    // Exception throwing is common; don't pay to calculate method name if it won't be logged.
+    if (_logger->IsEnabled(LogLevel::Debug))
+    {
+        FunctionID functionId = 0;
+        hr = _corProfilerInfo->DoStackSnapshot(
+            threadId,
+            ExceptionThrownStackSnapshotCallback,
+            COR_PRF_SNAPSHOT_INFO::COR_PRF_SNAPSHOT_DEFAULT,
+            &functionId,
+            nullptr,
+            0);
+
+        if (FAILED(hr) && hr != CORPROF_E_STACKSNAPSHOT_ABORTED)
+        {
+            LogErrorV(_T("DoStackSnapshot failed in function %s: 0x%08x"), to_tstring(__func__).c_str(), hr);
+            return hr;
+        }
+
+        tstring methodName;
+        IfFailLogRet(GetFullyQualifiedMethodName(functionId, methodName));
+        LogDebugV(_T("Exception thrown: %s"), methodName.c_str());
+    }
+
+    return S_OK;
+}
+
+HRESULT ExceptionTracker::ExceptionSearchCatcherFound(ThreadID threadId, FunctionID functionId)
+{
+    HRESULT hr = S_OK;
+
+    IfFailLogRet(_threadDataManager->SetExceptionCatcherFunction(threadId, functionId));
+
+    return S_OK;
+}
+
+HRESULT ExceptionTracker::ExceptionUnwindFunctionEnter(ThreadID threadId, FunctionID functionId)
+{
+    HRESULT hr = S_OK;
+
+    bool hasException = false;
+    FunctionID catcherFunctionId = ThreadData::NoFunctionId;
+    IfFailLogRet(_threadDataManager->GetException(threadId, &hasException, &catcherFunctionId));
+    IfFalseLogRet(hasException, E_UNEXPECTED);
+
+    if (ThreadData::NoFunctionId == catcherFunctionId)
+    {
+        tstring methodName;
+        IfFailLogRet(GetFullyQualifiedMethodName(functionId, methodName));
+        LogInformationV(_T("Exception unhandled: %s"), methodName.c_str());
+
+        // Future: Block thread until collection is initiated of the desired artifact.
+        // Possible serialization of some context of the exception and surrounding method
+        // information such as locals and parameters.
+    }
+    else if (functionId == catcherFunctionId)
+    {
+        IfFailLogRet(_threadDataManager->ClearException(threadId));
+
+        // Exception handling is common; don't pay to calculate method name if it won't be logged.
+        if (_logger->IsEnabled(LogLevel::Debug))
+        {
+            tstring methodName;
+            IfFailLogRet(GetFullyQualifiedMethodName(functionId, methodName));
+            LogDebugV(_T("Exception handled: %s"), methodName.c_str());
+        }
+    }
+
+    return S_OK;
+}
+
+HRESULT ExceptionTracker::ExceptionThrownStackSnapshotCallback(
+    FunctionID funcId,
+    UINT_PTR ip,
+    COR_PRF_FRAME_INFO frameInfo,
+    ULONG32 contextSize,
+    BYTE context[],
+    void* clientData)
+{
+    // Callback is only used to get the function ID of the top frame.
+    *static_cast<FunctionID*>(clientData) = funcId;
```

Comment:
If you use GetClassIDInfo2 in this callback with the frameinfo, you can probably get better information.

Reply:
This will also handle nested classes.

Reply:
Done